### PR TITLE
[ROCm] AMP selection process for AMD GPUs

### PR DESCRIPTION
**Pull request by stevenireeves**
Chooses to enable based on AMD GPU architecture, with current supported arches being gfx906 and gfx908. 
Utilizes the same Auto Mixed Precision tests as CUDA implementation. 

@cheshire @chsigg for review.

### CODE REVIEW

**Comment by stevenireeves** [comment]
The oneDNN failures may be related to the subject in https://github.com/tensorflow/tensorflow/pull/47646. Failures mark eigen::bfloat16.

**Comment by stevenireeves** [comment]
I reviewed the failures in the oneDNN CI build and I do not believe they are related to the changes I have made, as the GPU specific code and Intel MKL are separated by compiler directives.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 1]
```diff
@@ -1970,10 +1989,18 @@ int GetNumGPUs(const Cluster& cluster,
   int num_gpus = 0;
   for (const auto& device : devices) {
     const DeviceProperties& device_properties = device.second;
+#if GOOGLE_CUDA
```
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1992]: This seems to duplicate logic from `IsOnSuitableGPUArch`?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc [hunk 2]
```diff
@@ -102,6 +102,8 @@ class AutoMixedPrecisionTest : public GrapplerTest {
 #if GOOGLE_CUDA
     gpu_available_ =
         gpu_available_ && (num_gpus == GetNumAvailableGPUs(kMinGPUArch));
+#elif TENSORFLOW_USE_ROCM //Here we force Tensorflow to use the virtual GFX906
+    gpu_available_ = false;
```
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:106]: Why does ROCM require special logic here?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc [hunk 3]
```diff
@@ -1054,7 +1058,11 @@ TEST_F(AutoMixedPrecisionTest, BatchMatMul) {
 
   GraphView output_view(&output);
   EXPECT_EQ(output_view.GetNode("input")->attr().at("dtype").type(), DT_FLOAT);
+#if GOOGLE_CUDA
   if (GetCudaVersion(*virtual_cluster_.get()) >= 9010) {
```
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:1062]: Factor out to a function, e.g. `IsSupportedGPU`?

**Review by cheshire — CHANGES_REQUESTED** [review]

**Comment by cheshire** [comment]
Also adding Reed for sanity check on auto-mixed-precision stuff.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 4]
```diff
@@ -43,6 +43,23 @@ limitations under the License.
...
+
+#if TENSORFLOW_USE_ROCM
+// Returns true if the corresponding gfx arch string for the detected AMD GPU 
+// is in the list for FP16 supported compute. Returns false otherwise. 
+
+bool HasFastFP16Support(const DeviceProperties &props) {
+  absl::flat_hash_set<std::string> FP16SupportedDevices = {{"gfx906"}, {"gfx908"}};
+  std::string gcnArchName = props.environment().at("architecture");
+  std::vector<std::string> gpu_arch = absl::StrSplit(gcnArchName, ":");
+  if(gpu_arch.size()>0) {
```
Comment (cheshire) [inline tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:55]: return !gpu_arch.empty() && FP16... ?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc [hunk 5]
```diff
@@ -1160,7 +1175,11 @@ std::pair<int, int> GetDeviceGPUArch(
 }
 
 bool AutoMixedPrecisionImpl::IsOnSuitableGPUArch(const NodeDef& node) const {
+#ifndef TENSORFLOW_USE_ROCM
```
[the remark this answers is not in the document]
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I changed this to 
```
#if TENSORFLOW_USE_ROCM 
HasFP16Support(....)
#else 
....
#endif 
```

Is this what you mean? Or do you mean refactor in a way such that there's only one #if in the file?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 1 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1994]: This is counting the number of suitable GPUs in the cluster, in this regard I replicated the CUDA implementation using the rocm functionality. Here we are looping through the devices on a cluster and operating directly using their device properties, whereas `IsOnSuitableGPUArch` operates using nodes in the graph.

Comment (cheshire) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1994]: Then could we factor out the code which returns a bool from the device and call that from both?

File: tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc [hunk 6]
```diff
@@ -43,6 +43,16 @@ namespace tensorflow {
 namespace grappler {
 namespace {
 
+
+bool IsSupportedGPU()
+{
+#ifdef GOOGLE_CUDA
+    return GetCudaVersion(*virtual_cluster_.get()) >= 9010;
+#else
+    return True;
```
[the remark this answers is not in the document]
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision_test.cc:52]: Thanks for the catch.

File: tensorflow/core/grappler/optimizers/auto_mixed_precision.cc (hunk 1 above)
Comment (stevenireeves) [reply tensorflow/core/grappler/optimizers/auto_mixed_precision.cc]: I refactored `HasFastFP16Support` to do both GPU checks. This way the `#if`s are only in that function, cleans up both `IsOnSuitableGPU` and `GetNumGPUs`.

**Review by cheshire — APPROVED** [review]

**Comment by stevenireeves** [comment]
@cheshire I had a small mistake, which I fixed. Can we rerun the checks?

**Review by cheshire — APPROVED** [review]

**Review by reedwm — CHANGES_REQUESTED** [review]

**Review by reedwm — APPROVED** [review]

**Review by cheshire — APPROVED** [review]

**Comment by stevenireeves** [comment]
@cheshire @reedwm I accidentally pushed the wrong commit. The fix for the LinuxGPU failure was to move `IsOnSupportedGPU` after `GetCudaVersion` in `auto_mixed_precision_test.cc`. That is why `auto_mixed_precision_test.cc` couldn't compile the in the previous test. This should be the last review.

**Review by cheshire — APPROVED** [review]

**Pull request merged**